### PR TITLE
perf(repsel): open the array-element escape for `Ptr<Shape>` (#7034 §3)

### PR DESCRIPTION
**Pull request by proggeramlug**
`Ptr<Shape>` rule 2 disqualifies a local at any escape. #7034 §4 opened
`return`; this opens **`array/object element`** for the array half —
`rows.push(row)` and `for (const r of rows) r.field`.

`return` was easy because a return is a **terminator**: no use of the local can
follow it, so every access the pass licensed had already run while the object
was unaliased. An element escape is not. The object stays reachable through the
array for the rest of its life, so the containment region had to widen from
*one local* to *one local array and everything derived from it*, and the
array's own uses had to be bounded exactly as an object local's are.

## The rule (`collectors/ptr_shape_elements.rs`)

A region-local `A` is an **element-shape-proven array** of class `C` when all
of:

- **E1 provenance** — exactly one `Let { mutable: false, init: Array([]) }`,
  not boxed, not a module global. The literal must be **empty**: a non-empty
  one can carry elisions, whose slots read back as `undefined`.
- **E2 element provenance** — every write is `ArrayPush` of `new C(...)`,
  inline or via a local bound by one `Let { init: New { C } }` and pushed
  **exactly once**. No other mutator at all — no `pop`/`shift`/`splice`/
  `unshift`/`copyWithin`, no `IndexSet`, no `length` write. That is what makes
  `A` dense and monomorphic for its whole lifetime.
- **E3 array containment** — every other use is an in-bounds element read, a
  `.length` read, or `return A` (#7034 §4's terminator exemption, unchanged).
  Call argument, closure capture, reassignment, container element, unrecognised
  array method: all still disqualify.
- **E4 admissibility** — `C` passes the same `chain_admissible` gate rule 1
  applies to a `new C(...)` local, and the rule-5 module barrier is clear.
- **E5 in-bounds reads** — `A[i]` is licensed only where `i` is the induction
  variable of an enclosing `for (let i = 0; i < A.length; i++)`, written
  nowhere else in the region. **This conjunct is the whole difference between
  this pass and a wrong one.** Without it `A[i]` can be `undefined`, and a
  guard-free fixed-offset load masks a NaN-boxed `undefined` into a wild
  pointer. E3 admits no mutator that can shrink `A` and E2 makes it dense, so
  `0 <= i < A.length` at the read means `A[i]` is an own element of class `C`.

`for (const r of A)` desugars to exactly the E5 shape
(`lower/stmt_loops.rs::lazy_or_index_elem`), so the iterator form is covered by
the indexed proof rather than by a second one.

Then two halves in `ptr_shape.rs`: `A.push(row)` stops disqualifying `row`, and
`const r = A[i]` at a licensed site is rule-1 provenance of `new`-strength.

**Group integrity.** Every member of an element group — the pushed producers
plus the element-read locals — references objects the *other* members also
reach. One member failing rule 2 (`r.extra = 1`, a closure capture, an opaque
call) can transition the shape the others read guard-free, so the group is
all-or-nothing: `collect_shape_proven_ptr_locals` drops every member when any
one fails. No fixpoint is needed — dropping never admits a member.

**`numeric_fields` is not claimed** for group members. The numeric proof is an
exhaustive-reachable-store proof that containment makes possible *because no
alias exists*; a group has aliases by construction, and a sibling's
`r.score = "s"` (a declared field, so rule 2 permits it) downgrades the slot's
raw-f64 layout. Same stand-down as `proven_this.rs` and `ptr_shape_returns.rs`.
The shape proof alone still retires the whole guard diamond.

## What it buys, measured

On a build-then-consume kernel (40 000 records, produced with a local, then
read back both ways) the promoted function goes **0 → 3 selected, 3 consumed**,
and its emitted IR loses:

| symbol | base | after |
|---|---:|---:|
| `js_typed_feedback_class_field_get_guard` | 7 | **0** |
| `js_typed_feedback_record_fallback_call` | 7 | **0** |
| `js_object_get_field_by_name_f64` | 7 | **0** |
| all `js_*` calls | 72 | 51 |
| IR lines / blocks | 1426 / 127 | 1076 / 113 |

**Every other `js_*` call count is identical**, including all of
`js_shadow_slot_bind` (5), `js_write_barrier_root_nanbox` (5),
`js_write_barrier_slot` (1), `js_array_push_f64` (2), `js_gc_loop_safepoint`
(4) and the inline incremental-mark barrier sites (5). Nothing but guard
machinery went away.

## GC contract, verified in emitted IR

- The element locals get `js_shadow_slot_bind` in the entry block (slots 3 and
  4 of the probe). `TaPtr`'s callee-side no-bind shortcut is **not** copied —
  `GC_TYPE_OBJECT` is movable (#6990, #7019).
- Every access **re-derives** the raw pointer from that alloca:
  `load double, ptr %rN` → `and POINTER_MASK` → `inttoptr` → `gep +header` →
  `gep index` → `load`. The `for…of` local is reloaded 3× (3 field reads) and
  the indexed local 4× (4 access sites); nothing is cached across a safepoint.
- The store of the element into the bound slot is followed by the incremental
  mark-barrier check and `js_write_barrier_root_nanbox`.
- Write barriers on element stores are untouched: this pass changes no store
  lowering, and `js_array_push_f64` counts are identical between arms.
- The read side uses the `ptr_shape_get_number.plain` / `.coerce` pair — the
  2-instruction plain-finite check with a cold arm — because the group claims
  no numeric fields.

## What this does NOT reach, and the measurements that say so

**`batch.ts` is unchanged: 2 selected / 1 consumed, identical to `main`.** Both
of its element denials fail for reasons outside this rule:

- `buildRows`'s `const rows = []; …; return rows` never reaches the analysis —
  the **interprocedural deforestation pass** (`perry-transform/src/deforest`)
  has already rewritten it into a `__deforest_out` *parameter*, and a parameter
  array has no provenance. That transform fires on exactly the
  `const a = []; …push…; return a` producer shape this rule targets, which is a
  real coverage hole rather than an incidental one.
- `summarize`'s `byBucket` is passed as `rows.reduce(…)`'s seed — a call
  argument, so the array escapes (#7034 §1 territory).

**Dependency JS gets essentially nothing.** #7139 reported that ~103 candidates
its CJS barrier exemption freed were "immediately re-denied by rule 2", and
this position was picked on the assumption that those were element escapes.
They are not. Over **180** real `__esModule` CJS modules from
`real-apps/scriptc/node_modules`, compiled by a #7139-only arm and a
#7139+this-change arm (both 180/180), the 746 `Ptr<Shape>` candidates deny as:

| bucket | count |
|---|---:|
| rule 1 — allocation never bound to a local | **506** |
| rule 5 — module barrier still armed | 99 |
| rule 2 — bare reference | 130 |
| rule 2 — call argument | 5 |
| rule 2 — **array element** | **1** |
| rule 2 — closure capture / undeclared property | 1 / 1 |

Both arms are identical on every line. The rule-2 bare references are all
Perry's own `__cjs_module` wrapper local, and the 506 rule-1 denials break down
as constructor argument 182, statement 162, call argument 84, return 64, array
element 8, initializer 6. **The wall in dependency JS is rule 1 (allocations
never bound to a local), not containment.**

## Validation

- `cargo test -p perry-codegen --lib`: 429 passed (18 new).
- **Sabotage matrix, 15 conjuncts, each with a disjoint red set** — every guard
  deleted in turn, the suite re-run, and the failing tests recorded: push
  exemption, in-bounds read, both GC rooting obligations, group integrity,
  single-push, empty-literal seed, `const` array binding, shrinking mutators,
  indexed store, class agreement, index write count, bare array reference,
  closure capture, `.length` receiver. Control green in all 16 runs. Two
  weaknesses it caught and fixed: `a_local_pushed_into_two_arrays_is_not_exempt`
  passed on `HashMap` iteration order (now asserts the facts directly), and the
  closure-capture test denied through the body walk rather than the capture
  list.
- Census: new liveness fixture `fixture_ptr_shape_elements` with floors
  `ptr-shape 3 / ptr-shape-consumed 3`; no existing floor moved; the
  `PERRY_PTR_SHAPE_LOCALS=0` sabotage step in `repsel-census` now asserts the
  element fixture goes to zero too — without that, the whole analysis could
  stop issuing facts and every counter in the job would be unchanged.
- New `test_gap_repsel_ptr_shape_elements.ts`, registered in
  `test-parity/gc_repsel_corpus.txt`, byte-exact against Node 26.5.1: both read
  forms, 200 allocations *between* two field reads of the same element local,
  NaN/±Infinity/-0 written through one group member and read through another,
  and four arrays that must not be proven.

---

## Follow-ups filed from this work

- Deforestation hides the producer shape from this analysis (`buildRows`).
- Direct `A[i].field` reads are admitted by E3 but not promoted (consumption
  keys on `Expr::LocalGet`).
- Element reads bound to a `let`, non-empty array literals, and
  `sort`/`reverse`/`slice` on a proven array are all conservative refusals.
- Dependency JS's real wall is rule 1: 506 of 746 candidates are allocations
  never bound to a local.

## Summary by CodeRabbit

- **New Features**
  - Improved handling of objects stored in arrays, including indexed access and iteration.
  - Supports safe optimization for consistent object types and supported usage patterns.
  - Preserves correct behavior during garbage-collection movement and object mutations.
  - Added configurable, isolated Zig compiler caches for benchmark builds.

- **Bug Fixes**
  - Prevents unsafe optimizations for escaping arrays, mixed types, unsupported mutations, and out-of-bounds access.

- **Documentation**
  - Documented supported array-element behavior and limitations.

- **Tests**
  - Added regression, parity, and validation coverage for array access, iteration, GC movement, and edge cases.

### CODE REVIEW

File: crates/perry-codegen/src/collectors/ptr_shape_elements.rs [hunk 1]
```diff
@@ -0,0 +1,956 @@
...
+                    if self.root_of(*id).is_some() {
+                        self.disq(*id);
+                        return;
+                    }
+                }
+                if self.element_access_is_admissible(object.as_ref()) {
+                    return;
+                }
+                self.walk_expr(object);
+            }
```
[the remark this answers is not in the document]
Comment (proggeramlug) [reply crates/perry-codegen/src/collectors/ptr_shape_elements.rs:722]: **The hazard is real. The mechanism is already gone — verified by running your reproducer as a test first.**

`element_access_is_admissible` no longer exists. It was deleted in `816a5a3` (the commit this review was posted against), when I closed the adjacent unlicensed-element-read hole: **every** `IndexGet` occurrence outside the licensed `Stmt::Let` init now disqualifies the array, so a `PropertySet` whose receiver is an `IndexGet` reaches the `Expr::IndexGet` arm and kills the root.

I did not take that on faith. Your reproducer is now a permanent test, and it is **green on HEAD**:

- `a_property_store_through_an_element_denies_the_array` — your snippet verbatim in HIR form: `rows[i].extra = 1` in one bounded loop, `const r = rows[i]; r.x` in a second. Asserts both that `r` is not promoted **and** that `elements()` is empty, so it cannot pass by the read merely failing.
- `a_property_update_through_an_element_denies_the_array` — the same hazard via `A[i].f++`, which is a different HIR node (`PropertyUpdate`) and therefore a different arm. You were right to name both.

Sabotage confirms which guard carries them, disjointly:

| case | red set |
|---|---|
| `16_element_escape` (skip `disq` in the `IndexGet` arm) | both property tests + `an_element_passed_to_a_callee…` + `an_unlicensed_element_binding…` + `unbounded_element_read…` |
| `19_element_prop_store` (re-admit an `IndexGet` receiver in `PropertySet`) | `a_property_store_through_an_element_denies_the_array` |

I did not take the "record property names per root, validate after the class resolves" sketch: it would be needed only to *admit* declared-field element stores, which is strictly wider than today's rule. Disqualifying outright is the conservative direction, and #7151 tracks the widening — where your sketch is the right design, since the class genuinely is not known during the walk.

File: crates/perry-codegen/src/collectors/ptr_shape.rs [hunk 2]
```diff
@@ -507,12 +537,35 @@ pub(crate) fn collect_shape_proven_ptr_locals(
...
+                            &names,
+                            &depths,
+                            candidates.get(m).map(String::as_str),
+                            report::ESC_ELEMENT_GROUP,
+                        );
+                    }
+                }
+            }
+        }
+    }
```
[the remark this answers is not in the document]
Comment (proggeramlug) [reply crates/perry-codegen/src/collectors/ptr_shape.rs:590]: **Confirmed genuine and fixed.** Your reproducer went red on HEAD before any change, which is exactly what I wanted to see first:

```
test collectors::ptr_shape_elements::tests::group_integrity_drops_the_aliases_of_a_dropped_member ... FAILED
```

The diagnosis is exactly right: lines 533-537 hand the fact to every alias of a promoted root because an alias holds the same object, and the removal loop only removed the ids `group_members()` reports. So `const a = row` kept a guard-free `Ptr<Shape>` proof of a shape a sibling had just transitioned — all-or-nothing broken by the one construct that makes it necessary.

Applied your fix (alias closure ∪ members). The test is now `group_integrity_drops_the_aliases_of_a_dropped_member`, asserting all three of the offender, the member, and the alias are absent from `out`.

New sabotage case `ALIAS_CLOSURE` (removal reverted to `members` only) makes exactly that one test red, while `GROUP_INTEGRITY` (whole filter deleted) makes it and `one_member_failing_containment_voids_the_group` red — disjoint red sets, so the alias closure is independently load-bearing rather than incidentally covered by the filter it lives in.

Full matrix after the fix: 21 conjuncts, each with its own red set, control green at 55 tests.

File: crates/perry-codegen/src/collectors/ptr_shape_elements.rs [hunk 3]
```diff
@@ -0,0 +1,956 @@
...
+                    };
+                    self.pushes.entry(root).or_default().push(site);
+                    // Still walk for OTHER arrays nested in the value.
+                    if !matches!(value.as_ref(), Expr::LocalGet(_)) {
+                        self.walk_expr(value);
+                    }
+                    return;
+                }
+                self.walk_expr(value);
+            }
```
[the remark this answers is not in the document]
Comment (proggeramlug) [reply crates/perry-codegen/src/collectors/ptr_shape_elements.rs:751]: **Confirmed genuine and fixed with your patch, unchanged.** Red on HEAD before the change:

```
test collectors::ptr_shape_elements::tests::an_array_pushed_into_another_array_is_disqualified ... FAILED
```

Your reading of the control flow is precisely right: `PushValue::Other` disqualifies `array_id` (the outer array), but the `if !matches!(value.as_ref(), Expr::LocalGet(_))` guard below means the inner array's `LocalGet` never reaches the bare-reference arm, so it kept its facts while staying reachable and mutable through the outer array. `outer[0][0] = new Other()` is an `IndexSet` on an `IndexGet` — an expression neither walk tracks — so nothing else would have caught it either.

Applied verbatim, including your note about why it is safe: `disq(*v)` is a no-op for a producer local (`root_of` returns `None`), so the admitted `PushValue::Local` case is untouched — `pushed_local_is_promoted` and `a_clean_group_keeps_every_member` stay green.

Test: `an_array_pushed_into_another_array_is_disqualified`, your reproducer in HIR form, asserting both `element_read_class(r).is_none()` and that `r` is not promoted. Sabotage case `18_nested_array_push` makes exactly that one test red.

File: .github/workflows/test.yml [hunk 4]
```diff
@@ -1272,6 +1272,18 @@ jobs:
...
+          # real corpus workload promotes an element local -- so if it stopped
+          # issuing facts entirely, every assertion above would still pass and
+          # this job would stay green. Its own fixture is what makes that
+          # visible.
+          if ! printf '%s' "$out" | grep -q "fixture_ptr_shape_elements: ptr-shape-consumed promoted 0"; then
+            echo "::error::The array-element fixture kept its promotions with"
+            echo "::error::Ptr<Shape> disabled. Either the element analysis is not"
+            echo "::error::behind the knob, or the fixture stopped exercising it."
+            exit 1
+          fi
```
[the remark this answers is not in the document]
Comment (proggeramlug) [reply .github/workflows/test.yml:1300]: **Declining the change; the gap is real but the remedy is not mine to apply, and the deferral is deliberate.**

Two reasons, both project policy rather than preference:

1. **Branch protection is admin-only and the owner's call.** A PR cannot alter required contexts, so this is not something the change under review can fix.
2. **The deferral is documented and intentional.** `repsel-census` landed non-required on purpose (#7106), and the workflow says so at lines 1177-1180: *"DELIBERATELY NOT a required status check on its first landing — a gate that has never been green would block every open PR."* CLAUDE.md records the same rule generally: a new gate has never been green, so promote it after it has run, not on the commit that introduces it. Promoting it here would block every currently-open PR on a job none of them has ever run.

It is tracked as an open follow-through in `REPSEL-EXIT-NANBOX-PLAN.md` ("Promote `repsel-census` to a **required** status check once green"), and CLAUDE.md's own failure-mode (2) — *"not in branch protection's required contexts"* — is exactly the risk of leaving that undone. So the concern is right and recorded; only the timing is being deferred.

What this PR *can* do about the specific worry — that a failing array-element liveness gate merges unnoticed — it does: the assertion you flagged is paired with a sabotage step in the same job that requires the element fixture to go to **zero** under `PERRY_PTR_SHAPE_LOCALS=0`. Without that, the whole element analysis could stop issuing facts and every counter in the job would be unchanged (CLAUDE.md failure mode 4). Verified locally: `fixture_ptr_shape_elements: ptr-shape-consumed promoted 0, floor is 3`.

File: changelog.d/7149-ptr-shape-array-element-escape.md [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+  real coverage hole rather than an incidental one.
+- `summarize`'s `byBucket` is passed as `rows.reduce(…)`'s seed — a call
+  argument, so the array escapes (#7034 §1 territory).
+
+**Dependency JS gets essentially nothing.** #7139 reported that ~103 candidates
+its CJS barrier exemption freed were "immediately re-denied by rule 2", and
+this position was picked on the assumption that those were element escapes.
+They are not. Over **180** real `__esModule` CJS modules from
+`real-apps/scriptc/node_modules`, compiled by a #7139-only arm and a
+#7139+this-change arm (both 180/180), the 746 `Ptr<Shape>` candidates deny as:
```
[the remark this answers is not in the document]
Comment (proggeramlug) [reply changelog.d/7149-ptr-shape-array-element-escape.md]: Fixed. Reflowed so the sentence no longer begins with a hash:

> …compiled by a #7139-only arm and a combined (#7139 + this change) arm — both 180/180 — the 746 `Ptr<Shape>` candidates deny as:

MD018 clean; no `^#` line in the fragment now parses as an unintended ATX heading.

File: crates/perry-codegen/src/collectors/ptr_shape_elements_tests.rs [hunk 6]
```diff
@@ -0,0 +1,943 @@
...
+        was_unrolled: false,
+    };
+    let facts = facts_for(Vec::new(), vec![barrier]);
+    assert!(facts.has_shape_barrier_sites(), "fixture must be a barrier");
+    let stmts = vec![
+        let_arr(1, "rows"),
+        let_c(2, "row"),
+        push(1, Expr::LocalGet(2)),
+    ];
+    assert!(elements_with(&stmts, &classes, &facts).is_empty());
```
[the remark this answers is not in the document]
Comment (proggeramlug) [reply crates/perry-codegen/src/collectors/ptr_shape_elements_tests.rs:1144]: Taken. The invariant holds today — `pushed` is `retain`ed against `arrays`, `element_reads` only inserts for a proven root, and an empty `arrays` returns `default()` — but you are right that nothing enforces it and `is_empty()` gates every consumer, so a future edit populating one map without the other would fail silently.

Rather than widen `is_empty` (which would change the gate's meaning: `array_roots` is populated for alias tracking before any array is proven, so an all-maps check would report "non-empty" for a region with zero proven arrays), I made the invariant an assertion. `is_empty_covers_every_fact_map` checks **both** directions — a region whose array fails E3 has every map empty, and a proven region has none of them empty — with a `#[cfg(test)]` `debug_all_maps_empty()` accessor. Sabotage case `13_bare_array_ref` makes it red, so it is wired to the real behaviour rather than asserting a tautology.

**Comment by proggeramlug** [comment]
## Review round addressed — all 6 inline findings + all 4 nitpicks

**Every reproducer was added as a test before any fix**, so each finding had to prove itself red first. Result on the unfixed HEAD:

| finding | verdict |
|---|---|
| 🔴 group integrity drops no aliases | **RED — genuine**, fixed |
| 🟠 nested array push keeps inner facts | **RED — genuine**, fixed with your patch verbatim |
| 🔴 property store through `A[i]` admitted | **GREEN** — hazard real, mechanism already deleted in `816a5a3`; both reproducers kept as regression tests |
| 🟡 `is_empty` covers only `arrays` | **GREEN** — invariant holds; now asserted in both directions |
| 🟠 make `repsel-census` required | declined with rationale on the thread (admin-only + documented deliberate deferral) |
| 🟡 MD018 | fixed |

Nitpicks — all four taken, because three are the "test passes for the wrong reason" shape and one is the "two traversals drift" shape, both of which have bitten this campaign:

- `facts_for` now receives the test's **own** classes. A test that mutates its class and hands the mutation only to `chain_admissible`, while dispatch facts come from a pristine copy, can go green on a rule other than the one under test.
- The accessor fixture's getter was named `x` — also a declared field — so it could have passed on `FIELD_METHOD_AMBIGUITY` rather than the accessor rule. Renamed to `derived`.
- `element_read_seeds` reuses this module's one `walk_stmts` instead of a second copy of the traversal (−45 lines). Two traversals that must agree, drifting by one `Stmt` variant, silently withholds a seed on one side and a **disqualification** on the other. I also corrected `walk_stmts`' doc, which claimed it descends into closure bodies — it does not, and both callers are correct precisely because it doesn't.
- Added a direct unit test for the array **alias** path. `for (const r of A)` binds `const __arr_N = A` and indexes that, so the entire iterator read form runs through the alias edge, and only the end-to-end gap test covered it.

## Re-validation (rebased onto `8d3bc0e7c`, which includes #7139/#7140/#7141)

- `cargo test -p perry-codegen --lib`: **453 passed** (26 new).
- **Sabotage matrix: 21 conjuncts, each with a disjoint red set**, control green in all 22 runs. New cases `18_nested_array_push`, `19_element_prop_store`, `ALIAS_CLOSURE`.
- Census gate green, no floor moved; `PERRY_PTR_SHAPE_LOCALS=0` still takes the element fixture to 0.
- Oracle byte-exact vs Node 26.5.1 on the gap test, the census fixture, the kernel, and `batch.ts`.
- IR delta unchanged: 72 → 51 `js_*`, 1426 → 1076 lines, all 21 guard/feedback/by-name calls gone, **every GC-related count identical** (`js_shadow_slot_bind` 5/5, `js_write_barrier_root_nanbox` 5/5, `js_write_barrier_slot` 1/1, `js_array_push_f64` 2/2, `js_gc_loop_safepoint` 4/4).
- `gc_repsel_matrix --arms all --pressure 8` re-running against the fixed compiler; the previous run was PASS=447 UNVER=119 XFAIL=1 FAIL=0 with the new gap file PASS in all 21 arm columns and zero UNVER. Will post the new figure.

One note for the record: I also had to add the missing 7th argument to `collect_shape_proven_ptr_locals` in `cjs_scaffolding.rs`, which arrived with #7139 during the rebase.

**Comment by proggeramlug** [comment]
Matrix re-run against the fixed compiler is in: **`gc_repsel_matrix.sh --arms all --pressure 8` → PASS=447 UNVER=119 XFAIL=1 FAIL=0**, and `test_gap_repsel_ptr_shape_elements` is **PASS in all 21 arm columns with zero UNVER** — every arm, evacuating ones included, was live on it, so those greens are not the inert-arm kind (#6942/#6946/#6950). Identical to the pre-fix run, as expected: both review fixes are strictly narrowing.
